### PR TITLE
kernel: add manifest abstraction and a memory implementation

### LINKED ISSUE #164: Manifest abstraction to store kernel metadata
`Kernel` needs a metadata store for its metadata persistence. We have two choices:

- Add a general metadata abstraction in the same level with `Journal` and `Storage`
- Add a sub-module inside `Kernel` that only handles kernel specific metadata

For v0.2, we can choose the second solution for simplicity. We can still move to the first solution in the future if the second one is proofed to be inadequate.

### PR DESCRIPTION
**Pull request by huachaohuang**
Closes #164 

This PR uses protobuf to generate metadata so that we can serialize/deserialize it for storage and transport.

### CODE REVIEW

**Comment by huachaohuang** [comment]
@tisonkun @zojw This PR is ready. It is the prerequisite of #147 and #148 .

File: src/kernel/src/local/kernel.rs [hunk 1]
```diff
@@ -67,10 +78,14 @@ impl<S: Stream, B: Bucket> crate::Kernel for Kernel<S, B> {
 
     async fn apply_update(&self, update: KernelUpdate) -> Result<()> {
         let mut inner = self.inner.lock().await;
-        inner.last_sequence += 1;
+
+        let mut version = (*inner.current).clone();
```
Comment (zojw) [inline src/kernel/src/local/kernel.rs:82]: it reads inner.current in this tasks, but install new  inner.current in other task `handle_version_updates`?

do we have chance that "first flush's version hasn't installed but new flush coming?"

Comment (zojw) [reply src/kernel/src/local/kernel.rs:82]: emmm..maybe it has recorded by https://github.com/engula/engula/pull/143#discussion_r760111411

**Review by zojw — APPROVED** [review]
LGTM

File: src/kernel/src/local/kernel.rs (hunk 1 above)
Comment (zojw) [reply src/kernel/src/local/kernel.rs:82]: ignore me... I am wrong 🤣

**Review by tisonkun — APPROVED** [review]
LGTM. Although I'm a bit unsure whether we should create a layout:

```sh
local # abstract
mem # impl
```

at the same level. It's not a blocker for merging this PR.

**Comment by huachaohuang** [comment]
> LGTM. Although I'm a bit unsure whether we should create a layout:
> 
> ```shell
> local # abstract
> mem # impl
> ```

Yeah, the truth is, I am not sure about that either.

**Pull request merged**